earthaccess/search.py

```diff
@@ -160,14 +197,20 @@ def doi(self, doi: str) -> Type[CollectionQuery]:
 
         Parameters:
             doi: DOI of a datasets, e.g. 10.5067/AQR50-3Q7CS
+
+        Raises:
+            TypeError: if `doi` is not of type `str`
+
+        Returns:
+            self
         """
         if not isinstance(doi, str):
             raise TypeError("doi must be of type str")
 
         self.params["doi"] = doi
         return self
 
-    def instrument(self, instrument: str) -> Type[CollectionQuery]:
+    def instrument(self, instrument: str) -> Union[Self, Never]:
```

Comment:
The `Never` (or `NoReturn`, prior to Python 3.12) indicates that the function/method may throw an exception.

Reply:
https://docs.python.org/3/library/typing.html#typing.Never

Reply:
I'm not sure if I understand `Something | Never` annotation. My understanding is the reason to annotate `def foo() -> Never:` is to enable mypy to throw an error on `f = foo()`, since `foo` can't return. But if a function sometimes returns `Something` and sometimes returns `Never`, what additional checking does that enable?

My understanding is as a "bottom type", `Something | Never` is identical to `Something`!

I've read an argument that it serves to help document for humans that a possible outcome is an exception, but IMO that's a shortcoming of the type system that shouldn't be documented using the type system, since if you remove the possibility of an exception from the function and forget to also update the annotation, the annotation will be wrong and the typechecker won't know.

Reply:
Just to clarify, a return type of `Something | Never` does not indicate that it is possible for a function to return `Never`. It indicates that the function possibly never returns at all (typically indicative of the possibility of raising an exception _instead_ of returning a value).

However, I agree that Python's type system has shortcomings, so using `Never` really only serves to help the reader, but in thinking more about this, and seeing that an IDE may not even show `Never` in the function's return type when it pops up info when hovering your mouse pointer over a function call (at least VS Code doesn't show it, perhaps because it is a bottom type), then it would seem that this perhaps is not even helpful to the reader (unless directly reading the source code where the hint is written).

Further, given that a return type of `Something | Never` can be very confusing, since in all other cases, a union type indicates the possible types of return values, including `Never` doesn't actually makes sense since you can _never_ return `Never`.

Therefore, I'm going to drop the use of `Never`, but I'll keep the `Raises` sections that I added to the docstrings, since that indeed does show up in the little mouse-over popups in IDEs, and is helpful to include in generated docs.

> if you remove the possibility of an exception from the function and forget to also update the annotation, the annotation will be "wrong" (no longer serving its intended purpose) and the typechecker won't know.

As an aside, this isn't a particularly strong argument for removing `Never` because this issue is not unique to `Never`. There are currently plenty of places in the current codebase where mypy doesn't complain, but should, due to incorrect return type hints. In fact, this PR corrected many of them that mypy wasn't complaining about because of the very loose configuration of mypy, so this issue is also a matter of making sure we're also configuring the type checker effectively.

I added a bit of strictness with this PR, but not too much. Only enough to help weed out some of the incorrect hints scoped within the context of this PR. I plan to incrementally increase strictness so that we don't have to touch too many parts of the code at once.

If we don't improve our type hints, then we may be doing more harm than good since incorrect type hints are perhaps worse than no type hints, as with documentation.

Reply:
>Therefore, I'm going to drop the use of `Never`, but I'll keep the `Raises` sections that I added to the docstrings, since that indeed does show up in the little mouse-over popups in IDEs, and is helpful to include in generated docs.

That sounds great to me, thanks so much for looking in to IDE behavior around these annotations. These details of the developer experience aren't always paid attention to, and I love to see when they are! :heart: 


>I added a bit of strictness with this PR, but not too much. Only enough to help weed out some of the incorrect hints scoped within the context of this PR. I plan to incrementally increase strictness so that we don't have to touch too many parts of the code at once.
>
>If we don't improve our type hints, then we may be doing more harm than good since incorrect type hints are perhaps worse than no type hints, as with documentation.

:100: * :100: Couldn't agree more with the analysis and the approach. We need this and I really appreciate you're spearheading the work!